File: axiom/web/google_contacts.py
"""Marker search → preview → explicit selection → a draft/paused campaign."""
from __future__ import annotations

import secrets
import threading
import time
from pathlib import Path

from fastapi import APIRouter, Body, File, Form, UploadFile
from fastapi.responses import FileResponse, JSONResponse
from starlette.concurrency import run_in_threadpool

from db import database
from integrations import google_contacts as gc

router = APIRouter()
_previews: dict = {}
_lock = threading.Lock()
TTL = 30 * 60


def _error(exc):
    # OAuth/network exceptions can contain request details; do not echo them.
    msg = str(exc) if isinstance(exc, gc.ContactsError) else (
        "Не удалось прочитать Google Контакты. Проверьте подключение и повторите поиск; "
        "также можно загрузить экспорт Google CSV.")
    return JSONResponse({'error': msg}, status_code=400)


def _phone_index(conn):
    index = {}
    for row in conn.execute('SELECT id,phone,status,deleted_at,is_test,outreach_campaign_id FROM contacts WHERE phone IS NOT NULL'):
        key = gc.phone_key(row['phone'])
        if key:
            index.setdefault(key, []).append(dict(row))
    return index


def make_preview(rows, query):
    selected = gc.select_rows(rows, query)
    if len(selected) > 2000:
        raise gc.ContactsError("Найдено больше 2000 контактов. Уточните маркеры.")
    database.init_db()
    with database.get_conn() as conn:
        index = _phone_index(conn)
    for i, row in enumerate(selected):
        row['id'] = str(i)
        row['existing'] = {p: index.get(p, []) for p in row['phones']}
        row.pop('aliases', None)
    token = secrets.token_urlsafe(24)
    with _lock:
        now = time.time()
        for key in list(_previews):
            if _previews[key]['expires'] < now:
                del _previews[key]
        if len(_previews) >= 10:
            del _previews[next(iter(_previews))]
        _previews[token] = dict(expires=now + TTL, rows=selected, query=query)
    return dict(token=token, scanned=len(rows), matched=len(selected), items=selected)


@router.get('/assets/google-contacts.js')
def script():
    return FileResponse(Path(__file__).with_name('google-contacts.js'),
                        media_type='application/javascript', headers={'Cache-Control': 'no-cache'})


@router.get('/api/google-contacts/status')
def status():
    return {'connected': gc.TOKEN_FILE.exists()}


@router.post('/api/google-contacts/token')
async def upload_token(file: UploadFile = File(...)):
    try:
        raw = await file.read(65537)
        if len(raw) > 65536:
            raise gc.ContactsError('Файл токена слишком большой.')
        await run_in_threadpool(gc.save_token, raw.decode('utf-8-sig'))
        with _lock:
            _previews.clear()
        return {'ok': True}
    except Exception as exc:
        return _error(exc)


@router.post('/api/google-contacts/preview')
def preview(payload: dict = Body(...)):
    try:
        query = str(payload.get('markers') or '')
        gc.markers(query)  # Reject empty searches before reading the address book.
        return make_preview(gc.read_contacts(), query)
    except Exception as exc:
        return _error(exc)


@router.post('/api/google-contacts/csv-preview')
async def csv_preview(file: UploadFile = File(...), markers: str = Form(...)):
    try:
        gc.markers(markers)
        raw = await file.read(10 * 1024 * 1024 + 1)
        if len(raw) > 10 * 1024 * 1024:
            raise gc.ContactsError('Максимальный размер CSV — 10 МБ.')
        return await run_in_threadpool(lambda: make_preview(gc.from_csv(raw), markers))
    except Exception as exc:
        return _error(exc)


def import_selection(conn, snapshot, choices, campaign_id=None, campaign_name=''):
    """Caller holds a write transaction. Recheck duplicates and campaign state."""
    if not isinstance(choices, list) or not 1 <= len(choices) <= 500:
        raise gc.ContactsError('Выберите от 1 до 500 контактов.')
    rows = {r['id']: r for r in snapshot['rows']}
    selected, seen = [], set()
    for choice in choices:
        row = rows.get(str(choice.get('id')))
        phone = choice.get('phone', '')
        name = str(choice.get('person_name') or '').strip()
        if not row or phone not in row['phones']:
            raise gc.ContactsError('Выбор не соответствует предпросмотру. Повторите поиск.')
        if not name or len(name) > 100 or any(ord(c) < 32 for c in name):
            raise gc.ContactsError('Заполните имя для обращения у каждого выбранного контакта.')
        if any(m in gc.compact(name) for m in gc.markers(snapshot['query'])):
            raise gc.ContactsError('Уберите поисковый маркер из имени для обращения.')
        if phone not in seen:
            selected.append((row, phone, name))
            seen.add(phone)
    camp = None
    if campaign_id:
        camp = conn.execute('SELECT * FROM campaigns WHERE id=?', (int(campaign_id),)).fetchone()
        if not camp or camp['status'] not in ('draft', 'paused') or camp['archived']:
            raise gc.ContactsError('Выберите черновик или кампанию на паузе. Работающую сначала остановите в «Кампаниях».')
        if not (camp['audience_tag'] or '').strip():
            raise gc.ContactsError('У этой кампании не задана аудитория. Задайте отдельный тег в настройках кампании или создайте новый черновик.')
        if any(c in camp['audience_tag'] for c in ('%', '_')):
            raise gc.ContactsError('В теге кампании есть шаблонные символы % или _. Выберите отдельный буквенный тег.')
    elif not campaign_name.strip() or len(campaign_name) > 150:
        raise gc.ContactsError('Введите название нового черновика (до 150 символов).')
    index = _phone_index(conn)
    ready, skipped = [], []
    for row, phone, name in selected:
        matches = index.get(phone, [])
        reason = ''
        if len(matches) > 1:
            reason = 'В CRM несколько карточек с этим телефоном — требуется разбор дублей.'
        elif matches:
            old = matches[0]
            if old['deleted_at'] or old['is_test'] or old['status'] != 'new':
                reason = 'Контакт уже в работе, тестовый или в корзине; повторную рассылку не включали.'
            elif old['outreach_campaign_id'] and old['outreach_campaign_id'] != campaign_id:
                reason = 'Контакт закреплён за другой кампанией.'
        if reason:
            skipped.append(dict(name=row['name'], reason=reason))
        else:
            ready.append((row, phone, name, matches[0]['id'] if matches else None))
    if not ready:
        return dict(ok=True, added=0, existing=0, campaign_id=campaign_id, skipped=skipped)
    if camp is None:
        tag = 'Google-' + secrets.token_hex(8)
        campaign_id = conn.execute(
            "INSERT INTO campaigns (name,audience_tag,status,channel,auto_send,daily_limit) "
            "VALUES (?,?,'draft','telegram',0,5)", (campaign_name.strip(), tag)).lastrowid
    else:
        tag = camp['audience_tag']
    added = existing = 0
    ids = []
    for row, phone, name, cid in ready:
        if cid is None:
            cid = conn.execute(
                "INSERT INTO contacts (source,phone,name,person_name,email,tags,outreach_campaign_id) "
                "VALUES ('google_contacts',?,?,?,?,?,?)",
                (phone, row['name'], name, row.get('email') or None, tag, campaign_id)).lastrowid
            added += 1
        else:
            old = conn.execute('SELECT tags FROM contacts WHERE id=?', (cid,)).fetchone()
            tags = database._merge_tags(old['tags'], tag)
            # Preserve CRM name, conversation, source, and enrichment on repeat import.
            conn.execute("UPDATE contacts SET tags=?,outreach_campaign_id=?, "
                         "person_name=COALESCE(NULLIF(person_name,''),?),updated_at=datetime('now') WHERE id=?",
                         (tags, campaign_id, name, cid))
            existing += 1
        ids.append(cid)
    return dict(ok=True, added=added, existing=existing, campaign_id=campaign_id,
                contact_ids=ids, skipped=skipped)


@router.post('/api/google-contacts/import')
def import_contacts(payload: dict = Body(...)):
    try:
        with _lock:
            snapshot = _previews.get(str(payload.get('token') or ''))
        if not snapshot or snapshot['expires'] < time.time():
            raise gc.ContactsError('Предпросмотр устарел. Повторите поиск и выбор контактов.')
        database.init_db()
        with database.get_conn() as conn:
            conn.execute('BEGIN IMMEDIATE')
            result = import_selection(conn, snapshot, payload.get('selected'),
                                      int(payload['campaign_id']) if payload.get('campaign_id') else None,
                                      str(payload.get('campaign_name') or ''))
        return result
    except Exception as exc:
        return _error(exc)

File: axiom/web/google-contacts.js
/* Selective import. Only the server's preview can be submitted; no auto-send. */
window.showGoogleContacts = async function(panel) {
  if (panel.style.display !== 'none') { panel.style.display = 'none'; return; }
  panel.style.display = 'block';
  if (panel.dataset.ready) return;
  panel.dataset.ready = '1';
  const q = s => panel.querySelector(s);
  const escape = value => String(value ?? '').replace(/[&<>"']/g, c => ({'&':'&amp;','<':'&lt;','>':'&gt;','"':'&quot;',"'":'&#39;'}[c]));
  const request = async (url, options) => {
    const response = await fetch(url, options);
    const data = await response.json();
    if (!response.ok || data.error) throw new Error(data.error || 'Не удалось выполнить запрос.');
    return data;
  };
  const post = (url, body) => request(url, {method:'POST', headers:{'Content-Type':'application/json'}, body:JSON.stringify(body)});
  panel.innerHTML = `
    <div class="ttoolbar"><b>Выбрать контакты по маркерам в названии</b><span class="grow"></span><button data-close>Закрыть</button></div>
    <p class="hint">Введите один или несколько маркеров через запятую. Регистр и пробелы не важны. Например: эск2023, ск2025/2026.</p>
    <div class="ttoolbar" style="flex-wrap:wrap">
      <input data-markers value="эск2023, ск2025/2026" aria-label="Маркеры в названии" style="flex:1;min-width:230px">
      <select data-source aria-label="Источник поиска"><option value="google">Google Контакты</option><option value="csv">Загруженный CSV</option></select>
      <button class="primary" data-search>Найти</button>
      <button data-csv-button>Загрузить Google CSV</button>
      <input data-csv type="file" accept=".csv,text/csv" hidden>
    </div>
    <details style="margin:10px 0"><summary>Подключение Google Контактов</summary>
      <p data-connection class="hint">Проверяю подключение…</p>
      <p class="hint">Для прямого поиска нужен отдельный доступ на чтение контактов. Подключение календаря остаётся прежним.</p>
      <button data-token-button>Загрузить файл доступа</button><input data-token type="file" accept=".json" hidden>
      <p class="hint">Файл google_contacts_token.json получается при входе через tools/google_contacts_login.py на вашем компьютере.</p>
      <p class="hint">Без подключения: Google Контакты → Экспортировать → Google CSV. Загрузите файл здесь — в CRM попадут только отмеченные записи.</p>
    </details>
    <div data-message class="hint" role="status" style="margin:10px 0"></div>
    <div data-results></div>`;
  q('[data-close]').onclick = () => { panel.style.display = 'none'; };
  let savedFile = null, snapshot = null, busy = false;
  let campaigns = [];
  let connection = false;
  const message = text => { q('[data-message]').textContent = text; };
  const busySet = value => {
    busy = value;
    panel.querySelectorAll('button').forEach(b => { if (!b.hasAttribute('data-close')) b.disabled = value; });
  };
  try {
    const [status, list] = await Promise.all([request('/api/google-contacts/status'), request('/api/campaigns')]);
    connection = status.connected;
    campaigns = Array.isArray(list) ? list : [];
    q('[data-connection]').textContent = connection ? 'Файл доступа сохранён. Можно выполнить поиск.' : 'Прямой доступ к контактам ещё не подключён. Доступна загрузка Google CSV.';
  } catch (e) { message(e.message); }
  const render = data => {
    snapshot = data;
    const available = campaigns.filter(c => ['draft','paused'].includes(c.status) && !c.archived);
    message(`Проверено записей: ${data.scanned}. По маркерам найдено: ${data.matched}. Выберите нужные контакты и проверьте имена для обращения.`);
    const root = q('[data-results]');
    root.innerHTML = `
      <div class="ttoolbar"><button data-all>Выбрать доступные</button><button data-none>Снять выбор</button><span data-count class="hint">Выбрано: 0</span></div>
      <div style="overflow:auto;max-height:440px"><table style="width:100%"><thead><tr><th></th><th>Название в Google</th><th>Телефон</th><th>Имя для обращения</th><th>В AXIOM</th></tr></thead><tbody>
      ${data.items.map(r => `<tr data-row="${escape(r.id)}">
        <td><input type="checkbox" data-pick aria-label="Выбрать ${escape(r.name)}" ${r.phones.length?'':'disabled'}></td>
        <td>${escape(r.name)}<div class="hint">${escape(r.matched.join(', '))}</div></td>
        <td>${r.phones.length ? `<select data-phone aria-label="Телефон">${r.phones.map(p=>`<option value="${escape(p)}">${escape(p)}</option>`).join('')}</select>` : 'Нет подходящего телефона'}</td>
        <td><input data-person value="${escape(r.person_name)}" placeholder="Проверьте имя" aria-label="Имя для обращения" maxlength="100"></td>
        <td data-existing class="hint"></td></tr>`).join('')}
      </tbody></table></div>
      <div class="ttoolbar" style="margin-top:12px;flex-wrap:wrap">
        <label>Добавить в <select data-campaign aria-label="Кампания"><option value="">Новый черновик</option>${available.map(c=>`<option value="${c.id}">${escape(c.name)}</option>`).join('')}</select></label>
        <input data-campaign-name value="Клиенты — недвижимость" maxlength="150" aria-label="Название нового черновика">
        <button class="primary" data-import>Добавить выбранных</button>
      </div>
      <p class="hint">Контакты закрепляются только за выбранной кампанией. Добавление не запускает рассылку. Карточки с историей общения, тестовые и удалённые пропускаются.</p>`;
    const selectedCount = () => { q('[data-count]').textContent = 'Выбрано: ' + root.querySelectorAll('[data-pick]:checked').length; };
    root.querySelectorAll('[data-row]').forEach(tr => {
      const row = data.items.find(r=>r.id===tr.dataset.row);
      const phone = tr.querySelector('[data-phone]');
      const check = tr.querySelector('[data-pick]');
      const update = () => {
        const matches = row.existing[phone?.value] || [];
        const blocked = matches.length > 1 || matches.some(c=>c.deleted_at || c.is_test || c.status !== 'new');
        check.disabled = !phone || blocked;
        if (check.disabled) check.checked = false;
        tr.querySelector('[data-existing]').textContent = !matches.length ? 'Новый контакт' : matches.length > 1 ? 'Несколько дублей — пропуск' : blocked ? 'Уже в работе / исключён — пропуск' : 'Есть в CRM — без дубля';
        selectedCount();
      };
      if (phone) phone.onchange = update;
      check.onchange = selectedCount;
      update();
    });
    q('[data-all]').onclick = () => { root.querySelectorAll('[data-pick]:not(:disabled)').forEach(x=>{x.checked=true;}); selectedCount(); };
    q('[data-none]').onclick = () => { root.querySelectorAll('[data-pick]').forEach(x=>{x.checked=false;}); selectedCount(); };
    q('[data-campaign]').onchange = () => { q('[data-campaign-name]').hidden = !!q('[data-campaign]').value; };
    q('[data-import]').onclick = async () => {
      if (busy) return;
      const selected = [...root.querySelectorAll('[data-row]')].filter(tr=>tr.querySelector('[data-pick]').checked).map(tr=>({id:tr.dataset.row, phone:tr.querySelector('[data-phone]').value, person_name:tr.querySelector('[data-person]').value.trim()}));
      if (!selected.length) { message('Отметьте нужные контакты.'); return; }
      if (selected.some(r=>!r.person_name)) { message('Заполните имя для обращения у выбранных контактов.'); return; }
      busySet(true);
      try {
        const result = await post('/api/google-contacts/import', {token:snapshot.token, selected,
          campaign_id:q('[data-campaign]').value || null, campaign_name:q('[data-campaign-name]').value});
        message(`Добавлено: ${result.added}. Уже были в CRM: ${result.existing}. Пропущено: ${result.skipped.length}. Рассылка не запускалась.`);
        if (result.campaign_id) {
          q('[data-campaign]').value = String(result.campaign_id);
          if (!q('[data-campaign]').value) {
            const option = new Option(q('[data-campaign-name]').value, String(result.campaign_id));
            q('[data-campaign]').add(option); q('[data-campaign]').value = String(result.campaign_id);
          }
          q('[data-campaign-name]').hidden = true;
        }
        root.querySelectorAll('[data-pick]').forEach(x=>{x.checked=false;}); selectedCount();
        if (result.skipped.length) {
          const reasons = document.createElement('p'); reasons.className='hint';
          reasons.textContent = result.skipped.map(x=>`${x.name}: ${x.reason}`).join(' · '); root.appendChild(reasons);
        }
      } catch(e) { message(e.message); }
      finally { busySet(false); }
    };
  };
  const search = async useFile => {
    if (busy) return;
    const markers = q('[data-markers]').value.trim();
    if (!markers) { message('Введите маркеры.'); return; }
    if (!useFile && !connection) { q('details').open=true; message('Сначала подключите Google Контакты или загрузите Google CSV.'); return; }
    busySet(true); q('[data-results]').innerHTML=''; snapshot=null;
    message(useFile ? 'Ищу маркеры в файле…' : 'Читаю контакты Google и ищу маркеры во всех названиях…');
    try {
      let data;
      if (useFile) {
        const fd = new FormData(); fd.append('file', savedFile); fd.append('markers', markers);
        data = await request('/api/google-contacts/csv-preview', {method:'POST', body:fd});
      } else data = await post('/api/google-contacts/preview', {markers});
      render(data);
    } catch(e) { message(e.message); }
    finally { busySet(false); }
  };
  q('[data-search]').onclick = () => {
    if (q('[data-source]').value==='csv' && !savedFile) { q('[data-csv]').click(); return; }
    search(q('[data-source]').value==='csv');
  };
  q('[data-csv-button]').onclick = () => q('[data-csv]').click();
  q('[data-csv]').onchange = () => { savedFile=q('[data-csv]').files[0]; if(savedFile) { q('[data-source]').value='csv'; search(true); } };
  q('[data-token-button]').onclick = () => q('[data-token]').click();
  q('[data-token]').onchange = async () => {
    const file=q('[data-token]').files[0]; if(!file || busy)return;
    busySet(true);
    try {
      const fd=new FormData(); fd.append('file',file);
      await request('/api/google-contacts/token',{method:'POST',body:fd}); connection=true;
      q('[data-connection]').textContent='Доступ к контактам проверен. Можно выполнять поиск.';
      snapshot=null; q('[data-results]').innerHTML=''; message('Google Контакты подключены.');
    } catch(e) { message(e.message); }
    finally { busySet(false); }
  };
};
